k2/csrc/determinize.cc

```diff
@@ -459,15 +742,229 @@ class DetStateMap {
   };
 };
 
-void DeterminizeMax(const WfsaWithFbWeights &a, float beam, Fsa *b,
-                    std::vector<std::vector<int32_t>> *arc_map) {
-  // TODO(dpovey): use glog stuff.
-  assert(IsValid(a) && IsEpsilonFree(a) && IsTopSortedAndAcyclic(a));
-  if (a.arc_indexes.empty()) {
-    b->Clear();
-    return;
+/*
+  Convenience function that normalizes the state and outputs the arc for it.
+
+  Returns true if the state was newly added (not already present in
+  `state_map`).
+ */
+template <class TracebackState>
+bool NormalizeStateAndOutputArc(
+    DetState *state,
+    const WfsaWithFbWeights &wfsa_in,
+    float prune_cutoff,
+    vector<Arc> *arcs_out,
+    vector<float> *arc_weights_out,
+    vector<vector<typename TracebackState::DerivType> > *derivs_per_arc,
+    DetStateMap *state_map) {
+  float arc_weight;
+  std::vector<DerivType> deriv_info;
+  state->Normalize(wfsa_in, &arc_weight, &deriv_info);
+  int32_t next_state_id;
+  bool is_new_state = state_map->GetOutputState(state);
+  arcs_out->push_back({this->state_id, next_state_id, label});
```

Comment:
what does `this` point to? what's `next_state_id`, seems we can remove this function as there is a function `Normalize` in `DetState`?

Reply:
mm.. maybe this function should not exist but should be inline code, if it's called only once now.  this->state_id was the state id of the parent state, which is no longer in scope (was in ProcessArcs()).  next_state_id should be state->state_id.

Reply:
actually, possibly this function is never called and can be deleted.